### DISCUSSION TITLE
What does the `bins` parameter mean in NSF?

### DISCUSSION THREAD

**Discussion by Bill-Gots** [probabilists/zuko#43] [Q&A] [answered]
I checked the article _Neural Spine Flows_ and searched for the keyword `bins`. Then I got it' s related to some transforms like rantional-quadratic transform. But that article seemed so hard for me that I still don't understand why `bins` are related to the `shapes` parameter in `flows.MAF`. Could someone tell me some information about that?

**Comment by francois-rozet** [comment, marked as answer]
Hello, the number of bins is the number of *knots* minus one. The space between two knots is a bin. A spline is a piecewise function, meaning that each bin defines a different function, with different parameters. In the case of the rational-quadratic spline, each bin has 3 parameters: a width, a height and a slope. Hence, for the entire spline the parameters are 3 vectors of size `bins` (minus one for the slope due to boundary conditions). In comparison, for an affine transformation, the parameters are 2 scalars (tensor of shape `()`).

**Reply by Bill-Gots** [reply]
Thanks for reply! Just one more problem: how are the knots generated? Are they from neural networks whose input are randomly generated?

**Reply by francois-rozet** [reply]
The number of knots is chosen by the user. Their position is learned by the transformation during training. This is what the hyper network (`hyper`) returns.

**Reply by Bill-Gots** [reply]
Ok I got it. Thanks for your time!

**Comment by simonschnake** [comment]
The `flows.MAF` implementation creates a masked autoregressive flow (MAF). The architecture does not use a spline transform by default. There is an option to set `univariate=MonotonicRQSTransform` to use a *Neural Spline* architecture. As @francois-rozet explained, the number of bins is related to the number of knots in the spline transformation. Have a look at the graphic in the paper that should clarify the purpose of the knots.
